### PR TITLE
feat: use PerformanceObserver for gc stats

### LINKED ISSUE #5: GC stats
Consider adding garbage collection metrics.

### PR DESCRIPTION
**Pull request by acheronfail**
Fixes #5 

This uses the `PerformanceObserver` API to measure the default duration of garbage collections.
https://github.com/nodejs/diagnostics/blob/master/documentation/memory/step3/using_gc_traces.md?rgh-link-date=2020-06-17T22%3A50%3A46Z

The subscription calculates a rolling average over the `sampleInterval`.

### CODE REVIEW

**Comment by dnlup** [comment]
Thanks, @acheronfail , I appreciate it. Give me some time to test this, and I'll get back to you!

File: index.js [hunk 1]
```diff
@@ -23,12 +29,66 @@ function validate ({
...
+    if (key) {
+      let [count, average] = this[key]
+      if (average === undefined) {
+        this[key] = [1, gcEntry.duration]
+      } else {
+        count++
+        average -= average / count
+        average += gcEntry.duration / count
+        this[key] = [count, average]
+      }
```
Comment (dnlup) [inline index.js]: You used [this](https://en.wikipedia.org/wiki/Moving_average#Cumulative_moving_average) approach here, I like it.

**Comment by dnlup** [comment]
Could you rebase against the latest changes in `current`? There was a misconfiguration in GitHub actions that was preventing running tests on PRs :pray:

**Comment by acheronfail** [comment]
I'm not sure _how_ to deterministically trigger a `WeakCB` garbage collection cycle, so I've skipped checking for it in the tests.

Seeing as though it shares the same code as the other types, I think it's (probably) fine to leave it as is.

**Comment by acheronfail** [comment]
The tests run reliably on my computer (Linux), even if I run them 10 times in a row they still pass.
Not entirely sure how we can make these more reliable in CI.

**Comment by acheronfail** [comment]
~Okay, after running it in GitHub Actions several times in my fork, I've got it running reliably in CI. I just had to increase the duration of the garbage collection test. This way the test runs reliably since it's more likely that garbage collection occurs.~

I spoke too soon, it seems it still fails intermittently. :sigh: I'm not sure writing tests for the garbage collection is going to be easy, unless we re-architect it to allow us to mock the `PerformanceObserver` and send in dummy data during a test.

**Comment by dnlup** [comment]
> I'm not sure _how_ to deterministically trigger a `WeakCB` garbage collection cycle, so I've skipped checking for it in the tests.
> 
> Seeing as though it shares the same code as the other types, I think it's (probably) fine to leave it as is.

I agree.

> ~Okay, after running it in GitHub Actions several times in my fork, I've got it running reliably in CI. I just had to increase the duration of the garbage collection test. This way the test runs reliably since it's more likely that garbage collection occurs.~
> 
> I spoke too soon, it seems it still fails intermittently. :sigh: I'm not sure writing tests for the garbage collection is going to be easy, unless we re-architect it to allow us to mock the `PerformanceObserver` and send in dummy data during a test.

The latest tests seem to fail for something unrelated. I like the refactoring you have done so far, I would like to test the `PerformanceObserver` directly without mocking. For doing what you are suggesting the only way that comes to my mind atm is monkey patching the module `perf_hooks` to return a mock  or inject the mock as a dependency. Before taking that road I would like to try a few things.

I see errors in the custom `sampleInterval` check, mostly. Maybe it is something that could be related to the gc tests, but it seems that the tests on the gc were working with or without mocking. Let me try a few things locally. Maybe just being less strict on that timeout check could work.

I agree that testing this is not easy though, I encountered similar problems with the other metrics. We are going to find a solution.

**Comment by dnlup** [comment]
~~As a first small thing, let's put a debug message [here](https://github.com/dnlup/doc/blob/fae108f0a528015efeaf78ab42c60c247a36e0b9/test.js#L164):~~

```js
t.true(elapsed >= 2000 && elapsed < 2200, `elapsed: ${elapsed}`)
```
 
Edit: after digging a little more in the CI errors I am not really sure the problem is only in the interval check. Maybe the best thing we can do now is having this unit test on `GCStats`. I don't have a solution atm, but this feature works. If you could just address those nits I pointed in the previous comments we can proceed with this. This would allow me to publish a pre-release. I am confident something will come up to address these issues.

**Comment by acheronfail** [comment]
> If you could just address those nits I pointed in the previous comments we can proceed with this. This would allow me to publish a pre-release. I am confident something will come up to address these issues.

Sorry, I'm not entirely sure which nits you're referring to? Do you mean monkey-patching the `perf_hooks` module so we can mock out the `PerformanceObserver` class?

File: gc.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  constructor () {
+    this.reset()
+
+    // Start listening for garbage collection events
+    const gcObserver = new PerformanceObserver(this.observerCallback.bind(this))
+    gcObserver.observe({ entryTypes: ['gc'] })
+  }
+
+  observerCallback (list) {
+    for (const gcEntry of list.getEntriesByType('gc')) {
```
Comment (dnlup) [inline gc.js]: Here we can use `getEntries` because we subscribed to just gc events

File: gc.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      this.update(gcEntry)
+    }
+  }
+
+  reset () {
+    // These are set as a tuple of [count; rollingAverage]
+    this.major = [0, undefined]
```
Comment (dnlup) [inline gc.js]: Can the `average` be `0` as default?

File: gc.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    this.minor = [0, undefined]
+    this.incremental = [0, undefined]
+    this.weakCB = [0, undefined]
+  }
+
+  update (gcEntry) {
+    const key = GCStats.entryKindToKey(gcEntry)
+    if (key) {
+      let [count, average] = this[key]
+      if (average === undefined) {
```
Comment (dnlup) [inline gc.js]: If we set `average` to be `0` as default, we should check if `count === 0`

**Review by dnlup — CHANGES_REQUESTED** [review]

**Comment by dnlup** [comment]
> Sorry, I'm not entirely sure which nits you're referring to? Do you mean monkey-patching the perf_hooks module so we can mock out the PerformanceObserver class?

Sorry, I didn't submit my review correctly at the beginning so I think you didn't see my comments. I made a new one.

**Comment by acheronfail** [comment]
Ah, no worries! They're all done now. :+1:

**Review by dnlup — APPROVED** [review]

**Comment by dnlup** [comment]
Great work @acheronfail ! Thank you so much for working on this!

I'll publish a pre-release soon with this feature that you can try out by using the `next` tag when installing:

```bash
$ npm i @dnlup/doc@next
```

**Comment by dnlup** [comment]
Published https://github.com/dnlup/doc/releases/tag/v1.2.0-0

**Pull request merged**